### ISSUE TITLE
Implement zoom in Surge

### ISSUE THREAD

**Issue by baconpaul** [kurasu/surge#81]
I'm using this issue to discuss implementing zoom and my design for it.

In the audio unit host on MacOS, I now have zoom starting to work. Here's a screenshot of AU Lab with 3 different surges running each with a different zoom level.
 
<img width="1388" alt="screen shot 2018-12-14 at 5 59 05 pm" src="https://user-images.githubusercontent.com/13387561/50034379-7bc84800-ffca-11e8-9b3f-6d5c5890304c.png">

The zoom is user adjustable by a control which is next to about

<img width="339" alt="screen shot 2018-12-14 at 5 59 21 pm" src="https://user-images.githubusercontent.com/13387561/50034403-926e9f00-ffca-11e8-84b0-a10185d8a30d.png">

My diffs to do this are in baconpaul/surge on the branch zoomable-ui which is itself a branch off of au-cocoaui-build so none of this is remotely mergeable until #69 is merged into master.

There is one bug in my current zoom implementation, which is that the zoom factor doesn't restore properly after save.

But what I want to do on this thread, as well as ask people if they want to test my implementation if you are following along with my various forks (@esaruoho you know who I mean) is to discuss my implementation.

Basically I added to SurgeGUIEditor a factor called "zoomFactor" which is an integer. Those buttons increase or decrease it by a factor of 10. But each time they change, the gui editor also calls a ui callback which can be registered with the editor.

So in aulayer_cocoaui.mm I put this code:

```
        cont->setZoomCallback( [cont,self]() {
            AULOG::log( "In the Zoom Callback back in Cocoa to %d\n", cont->getZoomFactor() );
            ERect *vr;
            float zf = cont->getZoomFactor() / 100.0;
            if (cont->getRect(&vr))
            {
                NSRect newSize = NSMakeRect (0, 0,
                                             (int)( (vr->right - vr->left) * zf ),
                                             (int)( (vr->bottom - vr->top) * zf ) );
                AULOG::log( "Scalining display and unit square by %lf\n", zf );
                [self scaleUnitSquareToSize:NSMakeSize( zf / lastScale, zf / lastScale )];
                lastScale = zf;
                
                AULOG::log( "Resetting window size\n" );
                [self setFrame:newSize];
            }
            
        }
                              );
```

namely, when the scale changes I use the native toolkit to set the scale in the view. NSView can do this nicely due to the macOS resolution independent APIs, and I'm sure there is an equivalent context to scale in the VST implementation, just I don't know what it is. The nice thing about NSView is it scales all the events also so everything just works.

So what do people think? Is dropping a callback to the host to use the OS to do the scaling smart?

Also, note this scalability shows the difference between bitmap and vector assets. In the patch editor the text is drawn using native fonts so it is vector, and the other assets are bitmap, so look at this difference

<img width="505" alt="screen shot 2018-12-14 at 6 09 42 pm" src="https://user-images.githubusercontent.com/13387561/50034591-a2d34980-ffcb-11e8-93ca-ed36ed177a48.png">

but the nice thing is, from here, we could go about vectorizing bits of the UX step by step with non bitmap assets and get a bit of a hodge podge until we are done ... if that is we can render SVGs in VST which is a bit of work I keep contemplating but not actually starting!

Anyway comments welcome.

**Comment by cyanit** [comment]
Awesome! Can you change the color of the elements too? I made [this](https://github.com/cyanit/Surge-reskin) repo full of bitmaps (1x & 2x) and vectors, which could be used to give it a sleeker look, but only if the vector elements colors are adjusted from that orange to blue (exact color is in a colormap in the repo)

**Comment by baconpaul** [comment]
@cyanit is it possible when you export SVGs that you can export the text as paths?

Right now your svg has elements like this:

```
<text x="1052.32px" y="289px" style="font-family:'Roboto-Regular', 'Roboto';font-size:11px;fill:white;">Oscillat<tspan x="1088.79px " y="289px ">o</tspan>r</text>
```

but in inkscape at least you can replace the text with its path so you get only path entries and fill path entries.

Since, to use SVG, we are going to have to build a little svg parser and renderer inside vst land probably, the smaller a subset of SVG we deal with the better.

Hope this makes sense!

**Comment by baconpaul** [comment]
OK I fixed that last bug with some hosts where they resize window externally. I also put the zoom controls into a #if target_audiounit (sorry vst folks!) until someone implements in vst land.

But I think for AU on macOS I have zooming working reliably now. Of course, need to get #69 merged before I can do the pull request for this feature. But if you want to try it, the branch "zoomable-ui" in baconpaul/surge has everything you need.

**Comment by cyanit** [comment]
Btw, if the final goal is migrating from pixel graphics to SVGs, it might help if we split up the "grouped" PNGs into individual ones. This would have the benefit that the placement and logic are in place (and working).

 _SIDENOTE: Ideally, the positions are split to a (for non-programmers) readable file, so that people could reskin (and re-arrange) it easily down the road (some of you might know U-He's solution for this, I like the idea but they implemented in a way that if you want to completely reskin Zebra 2 you have to write up to 44 000 lines of script, which is NOT a graphic designer friendly solution)._

If somehow SVGs are a pain in the ass to implement, we could roll back on some scaling solution, that scales the individual elements. Since all the stuff would be vectorized anyway, exporting it into any resolution is not a problem.

**Comment by baconpaul** [comment]
Spliting the graphics makes it much much harder. If everything stays the same size with a vector layout then subbing in SVG looks doable. The reason there's all these "3 stacks in an image" is the UI right now just pans around the image based on state. Changing that is a total rewrite. Changing the thing around which we pan to be a vector thing may be an object plugin and a couple of tweaks.

But lets do nothing on the design side for a second until I sort of understand how it works. I grabbed an SVG from your repo which I"m using to experiment a bit.

**Comment by baconpaul** [comment]
So AU now has zoom but I'm going to leave this issue open for folks who want to implement it in VST or on Windows (@asimilon and @kzantow - that may be you!).

If you have any questions about how it works just drop them on this thread!

**Comment by kzantow** [comment]
I had a bit of a look at this, it appears that the VST code itself handles the macOS specific stuff... I'm looking in to how to change it without modifying the VST code directly, but haven't had much luck yet. I'm a complete noob when it comes to Objective-C and Cocoa framework. I'm sure it's doable, since other plugins can do this. Will report back with any progress I make. Also, I'd like to suggest a behavior tweak to this, where we store the last size (e.g. in `~/Library/Application Support/Surge/gui-scale.txt` or something similar) and use that when opening the plugin subsequent times. It gets to be very annoying to have to re-scale a plugin each time you open a new instance (try doing this with 50 instances o u-he Triple Cheese ;) ).

**Comment by baconpaul** [comment]
Where does the VST code do it? And can you tell me a way I can run the VST code on my Mac (I don’t even have a VST host running afaik).

We can probably play almost exactly the same trick if we get a reference to the nsview holding the component which must be in the vst lib.

**Comment by baconpaul** [comment]
Oh and yeah, I have zoomFactor hard coded to 200 in my user area since I always zoom anyway. I’d say more store a default size than last size though since if you have 5 surges running they will clobber each other. But having the default not be the teensy 100 is smart. That patch is host independent so happy for you to take a lead. Perhaps we just put the zoom into configuration.xml (but then we have to deal with updating it on your recursive push! So time for that version fix...)

**Comment by kzantow** [comment]
For a VST host, REAPER is quite good. Has a bit of a nag screen, but is only $60 to get rid of it. There's also the thing @esaruoho mentioned: https://github.com/tracktion/pluginval I just downloaded and built it, going to see how it works for testing VST/AU.

**Comment by kzantow** [comment]
As for version fix / recursive updates, I'd rather punt on that for a little while ;)

**Comment by baconpaul** [comment]
OK so I installed reaper and pluginval and both show me that "the vst doesn't work"!

pluginval was a little more useful. After doing a build and applying the diff to make package-vst3 use the executable from vst3 I

```
git diff package-vst3.sh
diff --git a/package-vst3.sh b/package-vst3.sh
index 7d93407..1297b84 100755
--- a/package-vst3.sh
+++ b/package-vst3.sh
@@ -5,7 +5,7 @@ RES_SRC_LOCATION="resources"
 PACKAGE_SRC_LOCATION="$RES_SRC_LOCATION/osx-vst2"
 BITMAP_SRC_LOCATION="$RES_SRC_LOCATION/bitmaps"
 BUNDLE_RES_SRC_LOCATION="$RES_SRC_LOCATION/osx-resources"
-EXEC_LOCATION="target/vst2/Release/Surge.dylib"
+EXEC_LOCATION="target/vst3/Release/Surge.dylib"
 
 # output configs

```

I see this

```
paul:~/dev/music/surge$ ~/Desktop/pluginval.app/Contents/MacOS/pluginval  --validate products/Surge.vst/
Random seed: 0x7146a27
-----------------------------------------------------------------
Starting test: CommandLineTests / Merge environment variables...
All tests completed successfully
-----------------------------------------------------------------
Starting test: CommandLineTests / Command line defaults...
All tests completed successfully
-----------------------------------------------------------------
Starting test: CommandLineTests / Command line parser...
All tests completed successfully
-----------------------------------------------------------------
Starting test: CommandLineTests / Command line random...
All tests completed successfully
-----------------------------------------------------------------
Starting test: AllocatorInterceptorTests / Ensure separate allocators are used for threads...
All tests completed successfully
-----------------------------------------------------------------
Starting test: AllocatorInterceptorTests / Ensure all allocations pass...
All tests completed successfully
-----------------------------------------------------------------
Starting test: AllocatorInterceptorTests / Ensure all allocations fail on MessageThread...
All tests completed successfully
-----------------------------------------------------------------
Starting test: AllocatorInterceptorTests / Ensure allocation violation status is reset...
All tests completed successfully
-----------------------------------------------------------------
Starting test: AllocatorInterceptorTests / Ensure all allocations pass on MessageThread, fail on background thread...
All tests completed successfully
-----------------------------------------------------------------
Starting test: AllocatorInterceptorTests / Ensure array allocations are caught...
All tests completed successfully
-----------------------------------------------------------------
Starting test: AllocatorInterceptorTests / Ensure allocations are thrown...
All tests completed successfully
pluginval v0.2.0 - JUCE v5.4.1
Started validating: products/Surge.vst/
Random seed: 0x48ad2a7
Validation started: 18 Dec 2018 2:39:04pm

Strictness level: 5
-----------------------------------------------------------------
Starting test: pluginval / Scan for known types: products/Surge.vst/...
Num types found: 0
!!! Test 1 failed: No types found
FAILED!!  1 test failed, out of a total of 1

Finished validating: products/Surge.vst/
*** FAILED: 1 TESTS
*** FAILED: 1 TESTS

2018-12-18 14:39:04.537 pluginval[25771:779500] *** Assertion failure in +[NSEvent startPeriodicEventsAfterDelay:withPeriod:], /BuildRoot/Library/Caches/com.apple.xbs/Sources/AppKit/AppKit-1671.10.106/AppKit.subproj/NSEvent.m:4345
2018-12-18 14:39:04.539 pluginval[25771:779500] *** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'Periodic events are already being generated'
*** First throw call stack:
(
	0   CoreFoundation                      0x00007fff51161e65 __exceptionPreprocess + 256
	1   libobjc.A.dylib                     0x00007fff7d1bd720 objc_exception_throw + 48
	2   CoreFoundation                      0x00007fff5117cab2 +[NSException raise:format:arguments:] + 98
	3   Foundation                          0x00007fff53520d1d -[NSAssertionHandler handleFailureInMethod:object:file:lineNumber:description:] + 194
	4   AppKit                              0x00007fff4ec3fea9 +[NSEvent startPeriodicEventsAfterDelay:withPeriod:] + 441
	5   pluginval                           0x0000000105474eb1 _ZN4juce12MessageQueue21runLoopSourceCallbackEPv + 129
	6   CoreFoundation                      0x00007fff510c7155 __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__ + 17
	7   CoreFoundation                      0x00007fff510c70fb __CFRunLoopDoSource0 + 108
	8   CoreFoundation                      0x00007fff510aab95 __CFRunLoopDoSources0 + 195
	9   CoreFoundation                      0x00007fff510aa13e __CFRunLoopRun + 1219
	10  CoreFoundation                      0x00007fff510a9a28 CFRunLoopRunSpecific + 463
	11  HIToolbox                           0x00007fff50342b35 RunCurrentEventLoopInMode + 293
	12  HIToolbox                           0x00007fff5034286b ReceiveNextEventCommon + 618
	13  HIToolbox                           0x00007fff503425e8 _BlockUntilNextEventMatchingListInModeWithFilter + 64
	14  AppKit                              0x00007fff4e5feeb7 _DPSNextEvent + 997
	15  AppKit                              0x00007fff4e5fdc56 -[NSApplication(NSEvent) _nextEventMatchingEventMask:untilDate:inMode:dequeue:] + 1362
	16  AppKit                              0x00007fff4e5f7cb9 -[NSApplication run] + 699
	17  pluginval                           0x00000001053b5f2b main + 347
	18  libdyld.dylib                       0x00007fff7e28c08d start + 1
	19  ???                                 0x0000000000000003 0x0 + 3
)
libc++abi.dylib: terminating with uncaught exception of type NSException
Abort trap: 6
```

Ooof. So "someone who wants to and knows how to debug VST3" seems like they are up next? Perhaps we should move this to a separate ticket for "VST3 works with pluginval" and then come back here once that is done?

**Comment by kzantow** [comment]
I just tried doing a VST2 build from master and already running into problems and it doesn't load in `pluginval`. I'll dig into this. Looks like there is already #60 for this; I'll have a look at it today in my spare time.

**Comment by baconpaul** [comment]
Hey OK I got a VST3 build working thanks to you and @asimilon which I could then install and run in a vst host I have running now (bitwig 2.4.3). So now I have a place where I can at least figure out the code which I need to intercept to implement zoom. Cool.

Thanks for those patches that made this work and install!

<img width="1488" alt="screen shot 2018-12-18 at 9 47 44 pm" src="https://user-images.githubusercontent.com/13387561/50195849-cd483e00-030e-11e9-9c9d-c43786230725.png">

**Comment by baconpaul** [comment]
And here's just a note on how I will take a swing at this, probably this week

the vst4gui uses an NSView subclass called NSViewFrame 

vst3sdk/vstgui4/vstgui/lib/platform/mac/cocoa/nsviewframe.mm

The editor seems to have a reference to this as "getPlatformFrame"

NSViewFrame* nsViewFrame = static_cast<NSViewFrame*> (frame->getPlatformFrame ());

so we should be able to have SurgeGUIEditor in VST mode register a zoom callback which, well, grabs that NSViewFrame then sends it the same messages. May require adding one objective c file to the VST3 build but that is easy enough. We already have them in src/mac.

And then the callback we write starts with a NSView cast then sends it the exact same messages that we send in au_cocoaui.mm in the AU.

I'll have a bit of time later this week to try that. If I do I'll post back here.

**Comment by kzantow** [comment]
@baconpaul great, I can pick that up and debug what's going on. I have some ideas to make the UI a little better, I think I mentioned somewhere before, will try to get some time in the next couple days to muck around with it.

**Comment by baconpaul** [comment]
Super

It resizes and scales properly butndoesnt do the origin mapping and window decoration handling. All the code is in the src/vst3 objective c file added on that branch.

I also have a bunch of work started on using svg in the ui rather than bitmap but the rebase on that failed. That said if your ideas include vector drawing and svg handling my “vectorized-draw” branch is a start. I’ll try and get that up to head plus diffs if I can.

Also @kurasu asked to make sure that we run any big changes to the ui by him before we merge as an FYI.

**Comment by kzantow** [comment]
Well, I don't have merge permissions, so no problem there ;)

**Comment by baconpaul** [comment]
OK so we are going to do zoom a “different way” in 2019 so I’m going to close this ticket and for now leave it implemented using the current scale only in the AU. The correct way (discussed in #23) will work in all hosts all platforms of course.
